### ISSUE TITLE
Converting from anyhow::Error

### ISSUE THREAD

**Issue by akshayknarayan** [yaahc/eyre#31]
Hi,
I am porting multiple crates from anyhow to eyre, so I have code that looks something like this:
  
```rust
fn foo() -> Result<(), anyhow::Error> {
  Ok(())
}

fn bar() -> Result<(), eyre::Report> {
  let f = foo().map_err(|e| eyre::Report::new(e))?;
  Ok(())
}
```

This, as expected, gives 
```
error[E0277]: the trait bound `anyhow::Error: std::error::Error` is not satisfied
  --> src/lib.rs:6:49
   |
6  |     let f = foo().map_err(|e| eyre::Report::new(e))?;
   |                                                 ^ the trait `std::error::Error` is not implemented for `anyhow::Error`
   |
  ::: /home/akshayn/.cargo/registry/src/github.com-1ecc6299db9ec823/eyre-0.6.0/src/error.rs:21:12
   |
21 |         E: StdError + Send + Sync + 'static,
   |            -------- required by this bound in `eyre::error::<impl eyre::Report>::new`

error: aborting due to previous error
```

Since `anyhow::Error` and `eyre::Report` are so similar, is there a better way to construct an `eyre::Report` mid-stack than this, which is sad and throws out the context:
```rust
fn foo() -> Result<(), anyhow::Error> {
  Ok(())
}

fn bar() -> Result<(), eyre::Report> {
  let f = foo().map_err(|e| eyre::eyre!("{:?}", e))?;
  Ok(())
}
```

Thanks for the help!

**Comment by yaahc** [comment]
the best way to convert between the two would be to first convert the anyhow::Error to a Box<dyn Error> and then convert that to an Eyre report with the macro. that way you'll preserve the chain of sources at least, though you'll still loss the backtrace that anyhow captured originally.

**Comment by akshayknarayan** [comment]
that makes sense, thanks!

**Comment by ppoliani** [comment]
> ```rust
> map_err(|err| eyre!(Box::new(err))
> ```

Alternatively you can wrap this inside a declarative macro

```rust
map_err!(foo());

#[macro_export]
macro_rules! map_err {
  ($fun:expr) => {
    $fun.map_err(|e| eyre!(Box::new(e)))?
  }
}
```

**Comment by pksunkara** [comment]
@yaahc Please let me know if you welcome a PR

**Comment by yaahc** [comment]
> @yaahc Please let me know if you welcome a PR

I've no objections to adding an anyhow_compat feature, PRs welcome

**Comment by thewh1teagle** [comment]
What's the status of this feature? 
I would like to use `?` on `anyhow` results on functions which return `eyre::Result`.

**Comment by sunshowers** [comment]
>  I would like to use ? on anyhow results on functions which return eyre::Result.

This is not permitted by Rust's orphan rules, unfortunately.

**Comment by sunshowers** [comment]
> Just for future readers:
> 
> ```rust
> foo().map_err(|err| eyre!(Box::new(err))
> ```

Ah sadly this will lose the source information, because it will produce a `Box<anyhow::Error>`. (I really don't like auto-deref specialization...).

What you want to do is:

```rust
    foo().map_err(|err| {
        eyre!(Box::<dyn std::error::Error + Send + Sync + 'static>::from(
            err
        ))
    });
```

Very wordy unfortunately. eyre can help here by providing a `new_boxed` method which would let type inference work.

**Comment by dtolnay** [comment]
> This is not permitted by Rust's orphan rules, unfortunately.

It is permitted. The `From` impls you would want are:

```rust
impl<E> From<E> for Report
where
    E: Into<anyhow::Error>,
    Result<(), E>: anyhow::Context<(), E>,
{
    fn from(error: E) -> Self {
        let anyhow_error: anyhow::Error = error.into();
        let boxed_error: Box<dyn StdError + Send + Sync + 'static> = anyhow_error.into();
        Report::from_boxed(boxed_error)
    }
}

impl From<Report> for anyhow::Error {
    fn from(report: Report) -> Self {
        let boxed_error: Box<dyn StdError + Send + Sync + 'static> = report.into();
        anyhow::Error::from_boxed(boxed_error)
    }
}
```

This would provide `?` in both directions.

**Comment by sunshowers** [comment]
@dtolnay oooooh that looks like a clever approach. Would that be a breaking change? (Or if gated by a feature, would it be non-additive? I guess `Into<anyhow::Error>` covers everything.)

**Comment by dtolnay** [comment]
It is additive with the existing impl.

I discussed this impl with eyre maintainers in June as part of https://github.com/eyre-rs/eyre/issues/177; not sure whether there's since then come up a reason not to implement it.

**Comment by yaahc** [comment]
I think the work has just been on hold for personal reasons mostly, tho I do recall a conversation about some issue w/ multiple layers of Box being introduced but I don't recall the details. I don't think we've abandoned the idea.
